### ISSUE TITLE
investigate: dense Forward applies RoPE before QK-norm (Qwen3 expects norm before RoPE)

### ISSUE THREAD

**Issue by pekkah** [pekkah/SharpInference#157]
## Summary

While generalizing the batched-trunk prefill to dense models (#156), I found that the **dense per-token `CudaForwardPass.Forward` applies RoPE _before_ per-head QK-norm**, whereas the Gemma path (`RunGemma4DeviceRegion`) applies **QK-norm before RoPE**.

```
// Forward (dense, ~CudaForwardPass.cs:940-958)
RoPE(q); RoPE(k);                       // RoPE first
if (_hasQkNorm && !UseL2QkNorm) { HeadNorm(q); HeadNorm(k); }   // then QK-norm

// RunGemma4DeviceRegion (~:1199-1223)
HeadNormQk(q,k);                        // QK-norm first
RoPE(q); RoPE(k);                       // then RoPE
```

RoPE does **not** commute with per-channel-weighted RMSNorm (RoPE preserves the per-pair L2 norm so the RMS scale is unchanged, but it mixes channels `i` and `i+d/2`, which carry different learned weights `w_i ≠ w_{i+d/2}`). The two orders produce materially different Q/K — in #156 the order mismatch alone caused a ~9-logit divergence on Qwen3-8B.

## Why this matters

The HF Qwen3 reference (and llama.cpp `build_qwen3`) apply **q_norm/k_norm before RoPE**. The dense per-token path here does the opposite. If that's a genuine deviation, dense Qwen3 (and any other weighted-QK-norm model: Qwen3-MoE, etc.) decode/prefill output may be subtly off vs the reference — even though it produces coherent text.

In #156 the batched path was deliberately made to **match the existing per-token oracle** (rope→norm for dense), so #156 introduces no behavior change — it faithfully reproduces whatever `Forward` already does. This issue is about whether `Forward` itself is correct.

## Investigation

- [ ] Confirm the actual order in HF `Qwen3Attention.forward` and llama.cpp `build_qwen3` (norm-then-rope expected).
- [ ] Check whether sharpi's top-1 parity claim vs llama.cpp b8585 on Qwen3-8B (README line ~19) was measured against this path — if it passed, reconcile how (e.g. weights near-uniform? parity tolerance? different code path?).
- [ ] Decide: if `Forward` is wrong, swap to norm→rope and re-validate parity; then update the #156 batched body to match (drop the per-family ordering branch — both would become norm→rope).
- [ ] Check the Vulkan `GpuForwardPass` and CPU `ForwardPass` for the same ordering, so a fix stays consistent across backends.

## Scope

Affects weighted (non-L2) QK-norm models on the dense paths. Gemma (norm→rope) and L2-QK-norm models are already correct. Not introduced by #156 — pre-existing.

Refs: #156 (where this was surfaced), branch `perf/cuda-dense-q4k-batched-prefill-156`.

**Comment by pekkah** [comment]
## Resolved — the dense GPU paths were genuinely inverted

Confirmed the bug and fixed it on branch `fix/cuda-dense-qknorm-rope-order-157` (commit 7e9a84a).

### The ordering is wrong, and it matters
RoPE does **not** commute with per-channel-weighted QK-norm. NEOX RoPE preserves each pair's L2 norm (so the RMS *scale* is unchanged), but it mixes channels `i` and `i+d/2`, which carry different learned `q_norm`/`k_norm` weights. So `RoPE(diag(w)·x) ≠ diag(w)·RoPE(x)` unless `w[i]==w[i+d/2]`. Qwen3 is exactly the affected config: `HasQkNorm=true`, `UseL2QkNorm=false`, `IsNeoxRope=true`.

### References all agree: norm → RoPE
- **HF Qwen3** `Qwen3Attention.forward`: `q_norm`/`k_norm` then `apply_rotary_pos_emb`.
- **llama.cpp `build_qwen3`** (`src/models/qwen3.cpp:85-101`): `build_norm(attn_q_norm)` → `ggml_rope_ext` (same for K).
- **sharpi CPU `ForwardPass`** (`ForwardPass.cs:645-654`): *"Qwen3 (weighted QK-norm): norm BEFORE RoPE"* — already correct.

The GPU dense paths did the opposite (RoPE → norm). The GDN paths (`CudaHybridGdnForwardPass`, `HybridGdnForwardPass`) and the Gemma paths were already norm → RoPE — only the generic *dense* paths were inverted.

### Why it stayed hidden (reconciling the README claim)
The CUDA-vs-CUDA batched-prefill oracles can't catch it: #156 deliberately built the batched path to **match the per-token loop**, so both were equally wrong and agreed with each other. The only independent reference was the CPU path. The README's "byte-identical 60-token greedy decode vs llama.cpp" is a **top-1 token-sequence** claim — robust to sub-argmax logit perturbation on clear-margin prompts, so it never tripped on the order bug. It is **not** evidence the order was correct; it's measuring a coarser thing than the bug perturbs.

### Numeric A/B (new cross-backend oracle, `Qwen3_8B_CudaForward_MatchesCpu_QkNormBeforeRope`)
CUDA dense `Forward` vs CPU `ForwardPass`, Qwen3-8B-Q4_K_M, identical prompt:
- **Pre-fix:** `maxAbs = 9.21` logits, top-5 overlap **3/5** → FAIL (matches the ~9-logit divergence #156 noted).
- **Post-fix:** argmax match, top-5 ≥ 4/5, `maxAbs < 4` → PASS.

### Fix (all dense paths, consistent with CPU/HF/llama.cpp)
- `CudaForwardPass`: `RunDeviceRegion`, `ForwardProfiled`, `GpuLayerBatchedTrunk` (dropped the per-family `_isGemma4Like` branch — both families are weighted → norm-first).
- `CudaHybridForwardPass`: per-token + `GpuLayerBatchedTrunk`.
- `GpuForwardPass` (Vulkan dense) and `HybridForwardPass` (Vulkan dense-hybrid).
- **L2 QK-norm (Llama-4) unchanged** — stays norm-AFTER-RoPE, RoPE-layers-only; the batched trunks already gate L2 out (`IsBatchedPrefillSupported`).

### Validation
- New cross-backend oracle (above).
- Qwen3 batched-prefill oracles **7/7** (incl. the bit-exact FlashOff one — batched/sequential stayed consistent since both were reordered together).
- Gemma4 CUDA **21/21**, dense-hybrid **5/5**, CUDA + Vulkan SnapKV forward **6/6**.
- One dense-hybrid oracle (`BatchedPrefill_ComputeRouting_ArgmaxStableVsMatvec_Coder`) relaxed from strict greedy-argmax to top-5-overlap + `maxAbs`: the *correct* output now lands on a **0.27-logit final-token tie** (362 @ 16.67 vs 4220 @ 16.40) that the argmax-stable fp16 GEMM tie-breaks differently than the byte-exact matvec — the same top-5 contract the sibling Qwen3 compute-routing oracles use.
- Qwen3-8B CUDA generation coherent end-to-end; decode 75.3 t/s (no regression). Full solution builds with 0 warnings.
